### PR TITLE
✨ New Source: Gainsight-px [Low Code CDK]

### LINKED ISSUE #3384: New Source: Gainsight
## Tell us about the new integration you’d like to have
I would like to pull data from [Gainsight.](https://support.gainsight.com/SFDC_Edition/API_and_Developer_Docs)

## Describe the context around this new integration
TBD -- pending feedback from user

## Describe the alternative you are considering or using

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368231736375) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by btkcodedev**
## What
*Developing new connector for the source: Gainsight*
*Closes #3384*

Resolves:
[🎉 New Source: Gainsight](https://github.com/airbytehq/airbyte/issues/3384)

## How
*Developed using (Configuration Based Source) low-code CDK*

## Recommended reading order
1. `spec.yaml`
2. `manifest.yaml`
3. `schemas/*`

## 🚨 User Impact 🚨
- No breaking changes, Just addition of new source
<details><summary><strong>New Connector: Source Gainsight</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [ ] Documentation updated
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

</details>

## Tests

<details><summary><strong>Integration & Acceptance</strong></summary>
Full Test Results

image

</details>

### CODE REVIEW

**Comment by btkcodedev** [comment]
Test results: Passed state, Ready for review

image

**Comment by btkcodedev** [comment]
Please ping via slack for getting the credentials

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/accounts.json [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      }
+    },
+    "createDate": {
+      "type": ["null", "string"]
+    },
+    "lastModifiedDate": {
+      "type": ["null", "string"]
+    },
+    "customAttributes": {
+      "type": ["null", "object"]
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/accounts.json:64]: missing the additionalProperties: true

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/feature.json [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    },
+    "propertyKey": {
+      "type": ["null", "string"]
+    },
+    "status": {
+      "type": ["null", "string"],
+      "enum": ["ACTIVATED", "DELETED"]
+    },
+    "featureLabels": {
+      "type": "object",
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/feature.json:28]:
```suggestion
      "type": ["null", "object"],
```

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/feature.json [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      "properties": {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/feature.json:29]: Add additionalProperties true to make backward compatibility

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/articles.json [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  "title": "Articles Schema",
+  "additionalProperties": true,
+  "type": ["null", "object"],
+  "properties": {
+    "id": {
+      "type": ["null", "string"]
+    },
+    "modifiedBy": {
+      "type": ["null", "object"],
+      "properties": {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/articles.json:12]: additionalProperties: true

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/articles.json [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        "name": {
+          "type": ["null", "string"]
+        }
+      }
+    },
+    "articleKCs": {
+      "type": ["null", "array"],
+      "items": {
+        "type": ["null", "object"],
+        "properties": {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/articles.json:25]: add additionalProperties true

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/kcbot.json [hunk 6]
```diff
@@ -0,0 +1,63 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "KCBot Schema",
+  "additionalProperties": true,
+  "type": "object",
+  "properties": {
+    "createdBy": {
+      "type": ["null", "object"],
+      "properties": {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/kcbot.json:9]: additionalProperties: true

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/kcbot.json [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      }
+    },
+    "createdDate": {
+      "type": ["null", "number"]
+    },
+    "description": {
+      "type": ["null", "string"]
+    },
+    "environments": {
+      "type": "array",
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/kcbot.json:25]:
```suggestion
      "type": ["null", "array"],
```

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/kcbot.json [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+      "items": {
+        "type": ["null", "string"]
+      }
+    },
+    "id": {
+      "type": ["null", "string"]
+    },
+    "modifiedBy": {
+      "type": ["null", "object"],
+      "properties": {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/kcbot.json:35]: additionalProperties: true

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/users.json [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    "sfdcContactId": {
+      "type": ["null", "string"]
+    },
+    "lastVisitedUserAgentData": {
+      "type": ["null", "array"],
+      "items": {
+        "globalUnsubscribe": {
+          "type": ["null", "boolean"]
+        }
+      }
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/users.json:96]: The item is an object please declare the `type` with the `globalUnsubscribe` fields

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/manifest.yaml [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    url_base: "https://api.aptrinsic.com/v1/"
+    http_method: "GET"
+    authenticator:
+      type: ApiKeyAuthenticator
+      header: "X-APTRINSIC-API-KEY"
+      api_token: "{{ config['api_key'] }}"
+
+  retriever:
+    type: SimpleRetriever
+    paginator:
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/manifest.yaml:27]: The connector allows to use pagination
image
Please implement it.

File: airbyte-integrations/connectors/source-gainsight-px/metadata.yaml [hunk 11]
```diff
@@ -0,0 +1,20 @@
+data:
+  allowedHosts:
+    hosts:
+      - TODO # Please change to the hostname of the source.
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gainsight-px/metadata.yaml:4]:
```suggestion
      - api.aptrinsic.com/v1
```

File: airbyte-integrations/connectors/source-gainsight-px/metadata.yaml [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  registries:
+    oss:
+      enabled: false
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gainsight-px/metadata.yaml:7]:
```suggestion
      enabled: true
```

File: airbyte-integrations/connectors/source-gainsight-px/metadata.yaml [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+  connectorSubtype: api
+  connectorType: source
+  definitionId: 0da3b186-8879-4e94-8738-55b48762f1e8
+  dockerImageTag: 0.1.0
+  dockerRepository: airbyte/source-gainsight-px
+  githubIssueLabel: source-gainsight-px
+  icon: gainsight-px.svg
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gainsight-px/metadata.yaml:14]: Please add the icon

**Review by marcosmarxm — DISMISSED** [review]
First review done! Thanks for the contribution.

File: airbyte-integrations/connectors/source-gainsight-px/metadata.yaml (hunk 13 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-gainsight-px/metadata.yaml:14]: Could you please provide the path for adding the icon, as the config-oss directory has recently changed

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/feature.json (hunk 3 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/feature.json:30]: Done

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/articles.json (hunk 4 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/articles.json:13]: Done

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/articles.json (hunk 5 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/articles.json:27]: Done

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/kcbot.json (hunk 6 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/kcbot.json:10]: Done

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/kcbot.json (hunk 8 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/schemas/kcbot.json:37]: Done

File: airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/manifest.yaml (hunk 10 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-gainsight-px/source_gainsight_px/manifest.yaml:27]: CursorPagination Implemented

File: airbyte-integrations/connectors/source-gainsight-px/metadata.yaml (hunk 13 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-gainsight-px/metadata.yaml:14]: Added in the same directory :)

**Comment by btkcodedev** [comment]
Comments resolved
Test state passed:

image

File: airbyte-integrations/connectors/source-gainsight-px/gainsight-px.svg
Comment (archangelic) [inline airbyte-integrations/connectors/source-gainsight-px/gainsight-px.svg:1]: This file needs to be renamed to `icon.svg`

File: airbyte-integrations/connectors/source-gainsight-px/metadata.yaml [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+  license: MIT
+  name: Gainsight Px
+  releaseDate: TODO
+  releaseStage: alpha
+  supportUrl: https://docs.airbyte.com/integrations/sources/gainsight-px
```
Comment (archangelic) [inline airbyte-integrations/connectors/source-gainsight-px/metadata.yaml:19]:
```suggestion
  documentationUrl: https://docs.airbyte.com/integrations/sources/gainsight-px
```

**Review by archangelic — CHANGES_REQUESTED** [review]

**Review by archangelic — APPROVED** [review]

**Comment by btkcodedev** [comment]
Thanks @archangelic I could see the CI is passing ✅

**Comment by archangelic** [comment]
Everything looks good to me. Once all the checks finished, I'll merge this!

**Pull request merged**